beginParsingChildren and finishParsingChildren should only exist on Element

rniwa opened the pull request:
#### 74df988dc5b55d9cfb4795e9b641df9fc032c343
<pre>
beginParsingChildren and finishParsingChildren should only exist on Element
<a href="https://bugs.webkit.org/show_bug.cgi?id=268213">https://bugs.webkit.org/show_bug.cgi?id=268213</a>

Reviewed by Anne van Kesteren and Darin Adler.

Only Element and ProcessingInstruction ever override finishParsingChildren.
ProcessingInstruction::finishParsingChildren is only there to unset m_createdByParser
flag but this function is only ever called by XMLDocumentParser::processingInstruction,
which can call ProcessingInstruction::setCreatedByParser instead. Furthermore, only
Element ever implemented beginParsingChildren.

This PR therefore moves beginParsingChildren and finishParsingChildren from Node
to Element and removes ProcessingInstruction::finishParsingChildren.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::finishParsingChildren):
(WebCore::Element::beginParsingChildren): Deleted.
* Source/WebCore/dom/Element.h:
(WebCore::Element::beginParsingChildren): Moved from cpp file.
* Source/WebCore/dom/Node.h:
(WebCore::Node::finishParsingChildren): Deleted.
(WebCore::Node::beginParsingChildren): Deleted.
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::finishParsingChildren): Deleted.
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::executeInsertTask):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseVoidElement):
* Source/WebCore/html/parser/HTMLElementStack.cpp:
(WebCore::HTMLElementStack::popAll):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLDocumentParser::endElementNs):
(WebCore::XMLDocumentParser::processingInstruction):

Canonical link: <a href="https://commits.webkit.org/273673@main">https://commits.webkit.org/273673@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f4990c0802164940cd3efd82f0f43d41386edce3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35998 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38716 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32379 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37219 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17387 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11973 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31099 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36552 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12653 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31960 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11084 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11105 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32143 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39964 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32719 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32472 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37049 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11266 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9179 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35133 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13009 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11773 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4701 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12104 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Node.h:
```diff
@@ -329,15 +329,6 @@ class Node : public EventTarget, public CanMakeCheckedPtr {
 
-    // Called by the parser when this element's close tag is reached,
-    // signaling that all child tags have been parsed and added.
-    // This is needed for <applet> and <object> elements, which can't lay themselves out
```
Note this comment is outdated since applet and object don't override these functions anymore.

annevk commented on Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
```diff
@@ -867,29 +867,30 @@ void XMLDocumentParser::endElementNs()
+    if (element)
+        element->finishParsingChildren();
 
     if (!scriptingContentIsAllowed(parserContentPolicy()) && is<Element>(*node) && isScriptElement(downcast<Element>(*node))) {
```
Might be nice to update this at the same time to take advantage of the new element variable.

rniwa replied:
Good point. Will fix.

annevk reviewed (approved):
This looks good modulo minor nit.

You might be interested in https://github.com/WebKit/WebKit/pull/23309 (still needs tests) where I discovered that `finishParsingChildren()` is not sufficient on its own for `script` elements to execute.

darinadler commented on Source/WebCore/dom/Element.h:
```diff
@@ -505,8 +505,9 @@ class Element : public ContainerNode {
     bool isFinishedParsingChildren() const { return isParsingChildrenFinished(); }
-    void finishParsingChildren() override;
-    void beginParsingChildren() final;
+    // Overriding these functions to make parsing a special case should be avoided if possible.
```
I don’t like this comment because it doesn't say why. Performance? Correctness? Inelegant solution to a problem best solved another way?

rniwa replied:
Hm... do you recall what was the original reasoning? The comment was added in https://github.com/WebKit/WebKit/commit/2ee64fe8e40dedcecc1b0948a5cc0a1da51714ac#diff-8d6a42a79982c64a568baee74bc315f946fdccaa399fa1ff272a668db4f85b2eR171.

As far as I know, special casing parser is bad because then elements won't respond to JS DOM mutations the same way but I'm wondering if there was some other reason for it.

rniwa replied:
Landing as is for now. Will update the comment later.

rniwa replied:
Updating the comment in https://github.com/WebKit/WebKit/pull/23448

rniwa committed:
beginParsingChildren and finishParsingChildren should only exist on Element
https://bugs.webkit.org/show_bug.cgi?id=268213

Reviewed by Anne van Kesteren and Darin Adler.

Only Element and ProcessingInstruction ever override finishParsingChildren.
ProcessingInstruction::finishParsingChildren is only there to unset m_createdByParser
flag but this function is only ever called by XMLDocumentParser::processingInstruction,
which can call ProcessingInstruction::setCreatedByParser instead. Furthermore, only
Element ever implemented beginParsingChildren.

This PR therefore moves beginParsingChildren and finishParsingChildren from Node
to Element and removes ProcessingInstruction::finishParsingChildren.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::finishParsingChildren):
(WebCore::Element::beginParsingChildren): Deleted.
* Source/WebCore/dom/Element.h:
(WebCore::Element::beginParsingChildren): Moved from cpp file.
* Source/WebCore/dom/Node.h:
(WebCore::Node::finishParsingChildren): Deleted.
(WebCore::Node::beginParsingChildren): Deleted.
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::finishParsingChildren): Deleted.
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::executeInsertTask):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseVoidElement):
* Source/WebCore/html/parser/HTMLElementStack.cpp:
(WebCore::HTMLElementStack::popAll):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLDocumentParser::endElementNs):
(WebCore::XMLDocumentParser::processingInstruction):

Canonical link: https://commits.webkit.org/273673@main
